ts/test/session/sending/MessageSender_test.ts

```diff
@@ -0,0 +1,176 @@
+        timestamp,
+        ttl,
+      });
+
+      const args = lokiMessageAPIStub.sendMessage.getCall(0).args;
+      expect(args[0]).to.equal(device);
+      expect(args[2]).to.equal(timestamp);
+      expect(args[3]).to.equal(ttl);
+    });
+
+    it('should correctly build the envelope', async () => {
+      stubEnvelopeType = SignalService.Envelope.Type.CIPHERTEXT;
+
+      // This test assumes the encryption stub returns the plainText passed into it.
+      const plainTextBuffer = crypto.randomBytes(10);
+      const timestamp = Date.now();
+
+      await MessageSender.send({
+        identifier: '1',
+        device: '0',
+        plainTextBuffer,
+        encryption: 0,
+        timestamp,
+        ttl: 1,
+      });
+
+      const data = lokiMessageAPIStub.sendMessage.getCall(0).args[1];
+      const webSocketMessage = SignalService.WebSocketMessage.decode(data);
+      expect(webSocketMessage.request?.body).to.not.equal(
+        undefined,
+        'Request body should not be undefined'
+      );
+      expect(webSocketMessage.request?.body).to.not.equal(
+        null,
+        'Request body should not be null'
+      );
+
+      const envelope = SignalService.Envelope.decode(
+        webSocketMessage.request?.body as Uint8Array
+      );
+      expect(envelope.type).to.equal(SignalService.Envelope.Type.CIPHERTEXT);
+      expect(envelope.source).to.equal(ourNumber);
+      expect(envelope.sourceDevice).to.equal(1);
+      expect(toNumber(envelope.timestamp)).to.equal(timestamp);
+      expect(envelope.content).to.deep.equal(plainTextBuffer);
+    });
+
+    describe('UNIDENTIFIED_SENDER', () => {
+      it('should set the envelope source to be empty', async () => {
+        stubEnvelopeType = SignalService.Envelope.Type.UNIDENTIFIED_SENDER;
+
+        // This test assumes the encryption stub returns the plainText passed into it.
+        const plainTextBuffer = crypto.randomBytes(10);
+        const timestamp = Date.now();
+
+        await MessageSender.send({
+          identifier: '1',
+          device: '0',
+          plainTextBuffer,
+          encryption: 0,
+          timestamp,
+          ttl: 1,
+        });
+
+        const data = lokiMessageAPIStub.sendMessage.getCall(0).args[1];
+        const webSocketMessage = SignalService.WebSocketMessage.decode(data);
+        expect(webSocketMessage.request?.body).to.not.equal(
+          undefined,
+          'Request body should not be undefined'
+        );
+        expect(webSocketMessage.request?.body).to.not.equal(
+          null,
+          'Request body should not be null'
+        );
+
+        const envelope = SignalService.Envelope.decode(
+          webSocketMessage.request?.body as Uint8Array
+        );
+        expect(envelope.type).to.equal(
+          SignalService.Envelope.Type.UNIDENTIFIED_SENDER
+        );
+        expect(envelope.source).to.equal(
+          '',
+          'envelope source should not exist in UNIDENTIFIED_SENDER'
+        );
```

Comment:
not sure if the label is correct here.
if the label is correct, you can use `expect(envelope).to.not.have.property(source)`

Reply:
or should it exist and be empty?

Reply:
it should exist but it needs to be empty